helper/metricsutil/gauge_process.go

```diff
@@ -0,0 +1,236 @@
+// so that collection processes do not all run at the time time.
+// If we knew all the procsses in advance, we could just schedule them
+// evenly, but a new one could be added per secret engine.
+func (p *GaugeCollectionProcess) delayStart() bool {
+	randomDelay := time.Duration(rand.Intn(int(p.currentInterval)))
+	// FIXME: a Timer might be better, but then we'd have to simulate
+	// one of those too?
+	delayTick := p.clock.NewTicker(randomDelay)
+	defer delayTick.Stop()
+
+	select {
+	case <-p.Stop:
+		return true
+	case <-delayTick.C:
+		break
+	}
+	return false
+}
+
+// resetTicker stops the old ticker and starts a new one at the current
+// interval setting.
+func (p *GaugeCollectionProcess) resetTicker() {
+	if p.ticker != nil {
+		p.ticker.Stop()
+	}
+	p.ticker = p.clock.NewTicker(p.currentInterval)
+}
+
+// collectAndFilterGauges executes the callback function,
+// limits the cardinality, and streams the results to the metrics sink.
+func (p *GaugeCollectionProcess) collectAndFilterGauges() {
+	// Run for only an allotted amount of time.
+	timeout := time.Duration(collectionBound * float64(p.currentInterval))
+	ctx, cancel := context.WithTimeout(context.Background(),
+		timeout)
+	defer cancel()
+
+	start := p.clock.Now()
+	values, err := p.collector(ctx)
+	end := p.clock.Now()
+	duration := end.Sub(start)
+
+	// Report how long it took to perform the operation.
+	p.sink.AddDurationWithLabels([]string{"metrics", "collection"},
+		duration,
+		p.labels)
+
+	// If over threshold, back off by doubling the measurement interval.
+	// Currently a restart is the only way to bring it back down.
+	threshold := time.Duration(collectionTarget * float64(p.currentInterval))
+	if duration > threshold {
+		p.logger.Warn("gauge collection time exceeded target", "target", threshold, "actual", duration, "id", p.labels)
+		p.currentInterval *= 2
+		p.resetTicker()
+	}
+
+	if err != nil {
+		p.logger.Error("error collecting gauge", "id", p.labels, "error", err)
+		return
+	}
+
+	// Filter to top N.
+	// This does not guarantee total cardinality is <= N, but it does slow things down
+	// a little if the cardinality *is* too high and the gauge needs to be disabled.
+	if len(values) > p.sink.MaxGaugeCardinality {
+		sort.Slice(values, func(a, b int) bool {
+			return values[a].Value > values[b].Value
+		})
+		values = values[:p.sink.MaxGaugeCardinality]
+	}
+
+	p.streamGaugesToSink(values)
+}
+
+func (p *GaugeCollectionProcess) streamGaugesToSink(values []GaugeLabelValues) {
+	// Dumping 500 metrics in one big chunk is somewhat unfriendly to UDP-based
+	// transport, and to the rest of the metrics trying to get through.
+	// Let's smooth things out over the course of a second.
+	// 1 second / 500 = 2 ms each, so we can send 25 per 50 milliseconds.
+	// That should be one or two packets.
+	sendTick := p.clock.NewTicker(50 * time.Millisecond)
+	batchSize := 25
+	for i, lv := range values {
+		if i > 0 && i%batchSize == 0 {
+			<-sendTick.C
```

Comment:
Should we also check if we're being shutdown here?

Reply:
Good idea, added.